packages/firestore/test/unit/specs/spec_test_runner.ts

```diff
@@ -1095,15 +1096,13 @@ abstract class TestRunner {
         undefined,
         'Expected active target not found: ' + JSON.stringify(expected)
       );
-      const actualTarget = actualTargets[targetId];
+      const { target: actualTarget, labels: actualLabels } =
+        actualTargets[targetId];
 
-      // TODO(mcg): populate the purpose of the target once it's possible to
-      // encode that in the spec tests. For now, hard-code that it's a listen
-      // despite the fact that it's not always the right value.
       let targetData = new TargetData(
         queryToTarget(parseQuery(expected.queries[0])),
         targetId,
-        TargetPurpose.Listen,
+        expected.targetPurpose,
```

Comment:
expected.targetPurpose either specified in the spec test file, or default to listen.